examples/common/pigweed/rpc_services/AttributeAccessor.h

```diff
@@ -0,0 +1,101 @@
+#include <app/AttributeReportBuilder.h>
+#include <app/AttributeValueDecoder.h>
+#include <app/AttributeValueEncoder.h>
+
+/**
+ * Callback class that clusters can implement in order to interpose custom
+ * attribute-handling logic.  An AttributeAccessor instance is associated
+ * with some specific cluster.  A single instance may be used for a specific
+ * endpoint or for all endpoints.
+ *
+ * Instances of AttributeAccessor that are registered via
+ * AttributeAccessorRegistry::Instance().Register will be consulted before taking the
+ * normal attribute access codepath and can use that codepath as a fallback if desired.
+ */
+namespace chip {
+namespace rpc {
+
+class AttributeAccessor
+{
+public:
+    /**
+     * aEndpointId can be Missing to indicate that this object is meant to be
+     * used with all endpoints.
+     */
+    AttributeAccessor(Optional<EndpointId> aEndpointId, ClusterId aClusterId) : mEndpointId(aEndpointId), mClusterId(aClusterId) {}
+    virtual ~AttributeAccessor() {}
+
+    /**
+     * Callback for reading attributes.
+     *
+     * @param [in] aPath indicates which exact data is being read.
+     * @param [in] aEncoder the AttributeValueEncoder to use for encoding the
+     *             data.
+     *
+     * The implementation can do one of three things:
+     *
+     * 1) Return a failure.  This is treated as a failed read and the error is
+     *    returned to the client, by converting it to a StatusIB.
+     * 2) Return success and attempt to encode data using aEncoder.  The data is
+     *    returned to the client.
+     * 3) Return success and not attempt to encode any data using aEncoder.  In
+     *    this case, Ember attribute access will happen for the read. This may
+     *    involve reading from the attribute store or external attribute
+     *    callbacks.
+     */
+    virtual ::pw::Status Read(const chip::app::ConcreteReadAttributePath & aPath, chip::app::AttributeValueEncoder & aEncoder) = 0;
+
+    /**
+     * Callback for writing attributes.
+     *
+     * @param [in] aPath indicates which exact data is being written.
+     * @param [in] aDecoder the AttributeValueDecoder to use for decoding the
+     *             data.
+     *
+     * The implementation can do one of three things:
+     *
+     * 1) Return a failure.  This is treated as a failed write and the error is
+     *    sent to the client, by converting it to a StatusIB.
+     * 2) Return success and attempt to decode from aDecoder.  This is
+     *    treated as a successful write.
+     * 3) Return success and not attempt to decode from aDecoder.  In
+     *    this case, Ember attribute access will happen for the write. This may
+     *    involve writing to the attribute store or external attribute
+     *    callbacks.
+     */
+    virtual ::pw::Status Write(const chip::app::ConcreteDataAttributePath & aPath, chip::app::AttributeValueDecoder & aDecoder)
+    {
+        return ::pw::Status::PermissionDenied();
```

Comment:
Permission denied is not the correct failure for "not found", Use ::pw::Status::NotFound